arbor/communication/communicator.cpp

```diff
@@ -0,0 +1,229 @@
+#include <utility>
+#include <vector>
+
+#include <arbor/assert.hpp>
+#include <arbor/common_types.hpp>
+#include <arbor/domain_decomposition.hpp>
+#include <arbor/recipe.hpp>
+#include <arbor/spike.hpp>
+
+#include "algorithms.hpp"
+#include "communication/gathered_vector.hpp"
+#include "connection.hpp"
+#include "distributed_context.hpp"
+#include "execution_context.hpp"
+#include "profile/profiler_macro.hpp"
+#include "threading/threading.hpp"
+#include "util/partition.hpp"
+#include "util/rangeutil.hpp"
+#include "util/span.hpp"
+
+#include "communication/communicator.hpp"
+
+namespace arb {
+
+communicator::communicator(const recipe& rec,
+                          const domain_decomposition& dom_dec,
+                          execution_context& ctx)
+{
+    distributed_ = ctx.distributed;
+    thread_pool_ = ctx.thread_pool;
+
+    num_domains_ = distributed_->size();
+    num_local_groups_ = dom_dec.groups.size();
+    num_local_cells_ = dom_dec.num_local_cells;
+
+    // For caching information about each cell
+    struct gid_info {
+        using connection_list = decltype(std::declval<recipe>().connections_on(0));
+        cell_gid_type gid;              // global identifier of cell
+        cell_size_type index_on_domain; // index of cell in this domain
+        connection_list conns;          // list of connections terminating at this cell
+        gid_info() = default;           // so we can in a std::vector
+        gid_info(cell_gid_type g, cell_size_type di, connection_list c):
+            gid(g), index_on_domain(di), conns(std::move(c)) {}
+    };
+
+    // Make a list of local gid with their group index and connections
+    //   -> gid_infos
+    // Count the number of local connections (i.e. connections terminating on this domain)
+    //   -> n_cons: scalar
+    // Calculate and store domain id of the presynaptic cell on each local connection
+    //   -> src_domains: array with one entry for every local connection
+    // Also the count of presynaptic sources from each domain
+    //   -> src_counts: array with one entry for each domain
+
+    // Record all the gid in a flat vector.
+    // These are used to map from local index to gid in the parallel loop
+    // that populates gid_infos.
+    std::vector<cell_gid_type> gids;
+    gids.reserve(num_local_cells_);
+    for (auto g: dom_dec.groups) {
+        util::append(gids, g.gids);
+    }
+    // Build the connection information for local cells in parallel.
+    std::vector<gid_info> gid_infos;
+    gid_infos.resize(num_local_cells_);
+    for (unsigned i=0; i<gids.size(); ++i) {
+        auto gid = gids[i];
+        gid_infos[i] = gid_info(gid, i, rec.connections_on(gid));
+    }
+    /*
+    threading::parallel_for::apply(0, gids.size(), thread_pool_.get(),
```

Comment:
Why remove the threaded parallel_for implementation here? Is it related to issue #789?

Reply:
Oh, this shouldn't have been removed. Good spot!